src/types.ts

```diff
@@ -573,20 +574,37 @@ export class Signature {
   cachedFunctionTarget: FunctionTarget | null = null;
   /** Respective function type. */
   type: Type;
+  /** The program that created this signature. */
+  program: Program;
 
   /** Constructs a new signature. */
   constructor(
+    program: Program,
     parameterTypes: Type[] | null = null,
     returnType: Type | null = null,
-    thisType: Type | null = null
+    thisType: Type | null = null,
   ) {
     this.parameterTypes = parameterTypes ? parameterTypes : [];
     this.parameterNames = null;
     this.requiredParameters = 0;
     this.returnType = returnType ? returnType : Type.void;
     this.thisType = thisType;
+    this.program = program;
     this.hasRest = false;
     this.type = Type.u32.asFunction(this);
+
+    let compare: Signature;
+    let signatureTypes = program.signatureTypes;
+    let length = signatureTypes.length;
+    for (let i = 0; i < length; i++) {
+      compare = signatureTypes[i];
+      if (this.isAssignableTo(compare)) {
```

Comment:
While this check works currently, the naming suggests that there can be multiple valid signatures. So I'm wondering if we should have an `equals` as well that we can use here, that `isAssignableTo` would be an alias of currently.

Reply:
Would it make sense to create an `equals<T>(value: T)` that returns `instanceof Signature && this.isAssignableTo(value)`?

Reply:
The `equals` I am proposing is a method `Signature#equals` that is then used here, to make this line of code more clear. For instance, in the future a signature with a trailing optional argument might be considered "assignable to" a signature without that trailing optional argument, but what we want here is actual uniqueness I think, hence `equals` is a better word. So I'm proposing:

1. Rename current `Signature#isAssignableTo` to `Signature#equals`
2. Introduce a new `Signature#isAssignableTo` that calls `equals` (for now)
3. Use `isAssignableTo` where something must be "assignable to" and `equals` where something must be "equal"

Reply:
Okay. That sounds reasonable. I'll work on this today.

Reply:
Should be all set!